README.md
# Cognitex
[![Build Status](https://travis-ci.org/Skoda091/cognitex.svg?branch=master)](https://travis-ci.org/Skoda091/cognitex)
[![hex.pm version](https://img.shields.io/badge/hex-v0.1.0-blue.svg)](https://hex.pm/packages/cognitex)

A simple library to handle user management over [AWS Cognito service](https://aws.amazon.com/cognito/).

The following functionality is covered;

* User registration
* Confirm registration (with the confirmation code received from AWS, by email)
* User authentication
* Fetch user data by an access token
* Fetch user data by username
* Change user password
* Update user attributes
* Reset a forgotten password

Online documentation is available [here](https://hexdocs.pm/cognitex/0.1.0/api-reference.html).

## Installation

To use `cognitex` with your projects, edit your `mix.exs` file to add it as a dependency:

```elixir
def deps do
  [
    {:cognitex, "~> 0.1.0"}
  ]
end
```

## Configuration

An example config might look like this:

```elixir
config :aws,
  key: "<AWS_ACCESS_KEY_ID>",
  secret: "<AWS_SECRET_ACCESS_KEY>",
  region: "<AWS_REGION>",
  client_id: "<AWS_CLIENT_ID>",
  user_pool_id: "<AWS_USER_POOL_ID>"
```

## Set up AWS Cognito with the correct configuration
First, we will set up a new AWS Cognito user pool with the correct configuration.

1. Visit your AWS console and go to the AWS Cognito service. Click on "Manage your User Pools" and click "Create a User Pool".
2. Specify a name for your pool and click "Review Defaults".
3. Optional: edit the password policy to remove some of the requirements. If you are just testing, using simple passwords will make it easier.
4. Click the "edit client" link. Specify a name for your app and be sure to *disable* the client secret and *enable* the `ADMIN_NO_SRP_AUTH` option.
5. Click "Create pool". Take note of the *Pool Id* at the top of the page and click on the apps page. Here, take note of the *App client id*.

## Usage

```elixir
iex> Cognitex.sign_up("john.smith@example.com", "Test123", name: "John", family_name: "Smith")
{:ok,
  %{
    "CodeDeliveryDetails" => %{
      "AttributeName" => "email",
      "DeliveryMedium" => "EMAIL",
      "Destination" => "j***@e***.co"
    },
    "UserConfirmed" => false,
    "UserSub" => "uuid"
  }
}

iex> Cognitex.sign_up("john.smith@example.com", "Test123", name: "John", family_name: "Smith")
{:error,
  %{
    message: "An account with the given email already exists.",
    status: "UsernameExistsException"
  }
}
```

Full documentation is available [here](https://hexdocs.pm/cognitex/0.1.0/api-reference.html).

### References

* Official Cognito [API reference](http://docs.aws.amazon.com/cognitoidentity/latest/APIReference/Welcome.html):

### Licensing, thanks

This library is released under the [MIT License](https://opensource.org/licenses/MIT).

It is based on:
* [cognito-phx](https://gitlab.com/azohra/cognito-phx/)
* [aws-elixir](https://github.com/aws-beam/aws-elixir)



CHANGELOG.md
# Changelog

## v0.1.0

* Initial release
* Basic functionality:
  * `Cognitex.sign_up/3` - Registers the user in the specified user pool and creates a user name, password, and user attributes defined in AWS Cognito service.
  * `Cognitex.confirm_sign_up/2` - Confirms registration of a user.
  * `Cognitex.admin_initiate_auth/2` - Authenticates registered user with credentials.
  * `Cognitex.get_user/1` - Gets the user attributes and metadata for a user by access token.
  * `Cognitex.admin_get_user/1` - Gets the specified user by username in a user pool as an administrator. Works on any user.
  * `Cognitex.change_password/3` - Changes the password for a specified user in a user pool.
  * `Cognitex.update_user_attributes/2` - Allows a user to update a specific attribute.
  * `Cognitex.forgot_password/1` - Calling this API causes a message to be sent to the end user with a confirmation code that is required to change the user's password.
  * `Cognitex.confirm_forgot_password/3` - Allows a user to enter a confirmation code to reset a forgotten password.


test/test_helper.exs
Mox.defmock(Services.CognitoMock, for: Services.Cognito)
Application.put_env(:cognitex, :cognito, Services.CognitoMock)

ExUnit.start(exclude: [:skip])


test/cognitex_test.exs
defmodule CognitexTest do
  use ExUnit.Case
  import Mox
  alias Services.CognitoMock

  setup do
    Application.put_env(:aws, :client_id, "client_id")
    Application.put_env(:aws, :user_pool_id, "user_pool_id")
    :ok
  end

  describe "sign_up/3" do
    setup :verify_on_exit!

    setup do
      {:ok,
       %{
         username: "john.smith@example.com",
         password: "test123",
         attrs: [name: "John", family_name: "Smith"]
       }}
    end

    test "should register user with correct data", %{
      username: username,
      password: password,
      attrs: [name: name, family_name: family_name] = attrs
    } do
      cognito_input = %{
        ClientId: "client_id",
        Password: password,
        UserAttributes: [
          %{Name: "name", Value: name},
          %{Name: "family_name", Value: family_name}
        ],
        UserPoolId: "user_pool_id",
        Username: username
      }

      cognito_output = %{
        "CodeDeliveryDetails_duap" => %{
          "AttributeName" => "email",
          "DeliveryMedium" => "EMAIL",
          "Destination" => "j***@e***.co"
        },
        "UserConfirmed" => false,
        "UserSub" => "uuid"
      }

      expect(CognitoMock, :sign_up, fn cognito_input ->
        send(self(), {:sign_up, cognito_input})
        {:ok, cognito_output, %{}}
      end)

      assert {:ok, ^cognito_output} = Cognitex.sign_up(username, password, attrs)

      assert_received {:sign_up, ^cognito_input}
    end

    test "should fail when user already exists for provided username and return error", %{
      username: username,
      password: password,
      attrs: [name: name, family_name: family_name] = attrs
    } do
      cognito_input = %{
        ClientId: "client_id",
        Password: password,
        UserAttributes: [
          %{Name: "name", Value: name},
          %{Name: "family_name", Value: family_name}
        ],
        UserPoolId: "user_pool_id",
        Username: username
      }

      cognito_status = "UsernameExistsException"
      cognito_message = "An account with the given email already exists."

      expect(CognitoMock, :sign_up, fn cognito_input ->
        send(self(), {:sign_up, cognito_input})
        {:error, {cognito_status, cognito_message}}
      end)

      assert {:error, %{message: ^cognito_message, status: ^cognito_status}} =
               Cognitex.sign_up(username, password, attrs)

      assert_received {:sign_up, ^cognito_input}
    end
  end

  describe "confirm/2" do
    setup :verify_on_exit!

    setup do
      {:ok, %{username: "john.smith@example.com", confirmation_code: "065045"}}
    end

    test "should confirm user registration with a valid confirmation code", %{
      username: username,
      confirmation_code: confirmation_code
    } do
      cognito_input = %{
        ClientId: "client_id",
        Username: username,
        ConfirmationCode: confirmation_code
      }

      expect(CognitoMock, :confirm_sign_up, fn cognito_input ->
        send(self(), {:confirm_sign_up, cognito_input})
        {:ok, %{}, ""}
      end)

      assert {:ok, %{}} = Cognitex.confirm(username, confirmation_code)

      assert_received {:confirm_sign_up, ^cognito_input}
    end

    test "should fail when user is already confirmed and return error", %{
      username: username,
      confirmation_code: confirmation_code
    } do
      cognito_input = %{
        ClientId: "client_id",
        Username: username,
        ConfirmationCode: confirmation_code
      }

      cognito_status = "NotAuthorizedException"
      cognito_message = "User cannot be confirm. Current status is CONFIRMED"

      expect(CognitoMock, :confirm_sign_up, fn cognito_input ->
        send(self(), {:confirm_sign_up, cognito_input})
        {:error, {cognito_status, cognito_message}}
      end)

      assert {:error, %{message: ^cognito_message, status: ^cognito_status}} =
               Cognitex.confirm(username, confirmation_code)

      assert_received {:confirm_sign_up, ^cognito_input}
    end
  end

  describe "authenticate/2" do
    setup :verify_on_exit!

    setup do
      {:ok, %{username: "john.smith@example.com", password: "test123"}}
    end

    test "should authenticate user with a valid credentioals", %{
      username: username,
      password: password
    } do
      cognito_input = %{
        AuthFlow: "ADMIN_NO_SRP_AUTH",
        AuthParameters: %{PASSWORD: password, USERNAME: username},
        ClientId: "client_id",
        UserPoolId: "user_pool_id"
      }

      cognito_output = %{
        "AuthenticationResult" => %{
          "AccessToken" => "access_token",
          "ExpiresIn" => 3600,
          "IdToken" => "id_token",
          "RefreshToken" => "refresh_token",
          "TokenType" => "Bearer"
        },
        "ChallengeParameters" => %{}
      }

      expect(CognitoMock, :admin_initiate_auth, fn cognito_input ->
        send(self(), {:admin_initiate_auth, cognito_input})
        {:ok, cognito_output, ""}
      end)

      assert {:ok, ^cognito_output} = Cognitex.authenticate(username, password)

      assert_received {:admin_initiate_auth, ^cognito_input}
    end

    test "should fail when incorrect credentials are provided and return error", %{
      username: username,
      password: password
    } do
      cognito_input = %{
        AuthFlow: "ADMIN_NO_SRP_AUTH",
        AuthParameters: %{PASSWORD: password, USERNAME: username},
        ClientId: "client_id",
        UserPoolId: "user_pool_id"
      }

      cognito_status = "NotAuthorizedException"
      cognito_message = "Incorrect username or password."

      expect(CognitoMock, :admin_initiate_auth, fn cognito_input ->
        send(self(), {:admin_initiate_auth, cognito_input})
        {:error, {cognito_status, cognito_message}}
      end)

      assert {:error, %{message: ^cognito_message, status: ^cognito_status}} =
               Cognitex.authenticate(username, password)

      assert_received {:admin_initiate_auth, ^cognito_input}
    end
  end

  describe "get_user/1" do
    setup :verify_on_exit!

    setup do
      {:ok,
       %{
         token: "token",
         email: "john.smith@example.com",
         name: "John",
         family_name: "Smith",
         sub: "sub_id"
       }}
    end

    test "should return user data with provided token", %{
      token: token,
      email: email,
      name: name,
      family_name: family_name,
      sub: sub
    } do
      cognito_input = %{
        AccessToken: token
      }

      cognito_output = %{
        "UserAttributes" => [
          %{"Name" => "sub", "Value" => sub},
          %{"Name" => "email_verified", "Value" => "true"},
          %{"Name" => "name", "Value" => name},
          %{"Name" => "family_name", "Value" => family_name},
          %{"Name" => "email", "Value" => email}
        ],
        "Username" => sub
      }

      expect(CognitoMock, :get_user, fn cognito_input ->
        send(self(), {:get_user, cognito_input})
        {:ok, cognito_output, ""}
      end)

      assert {:ok, user_attributes} = Cognitex.get_user(token)

      assert %{
               user_attributes: %{
                 email: ^email,
                 email_verified: "true",
                 family_name: ^family_name,
                 name: ^name,
                 sub: ^sub
               }
             } = user_attributes

      assert_received {:get_user, ^cognito_input}
    end

    test "should fail when incorrect token provided and return error", %{token: token} do
      cognito_input = %{
        AccessToken: token
      }

      cognito_status = "NotAuthorizedException"
      cognito_message = "Could not verify signature for Access Token"

      expect(CognitoMock, :get_user, fn cognito_input ->
        send(self(), {:get_user, cognito_input})
        {:error, {cognito_status, cognito_message}}
      end)

      assert {:error, %{message: ^cognito_message, status: ^cognito_status}} =
               Cognitex.get_user(token)

      assert_received {:get_user, ^cognito_input}
    end
  end

  describe "admin_get_user/1" do
    setup :verify_on_exit!

    setup do
      {:ok,
       %{username: "john.smith@example.com", name: "John", family_name: "Smith", sub: "sub_id"}}
    end

    test "should return user data with provided username", %{
      username: username,
      name: name,
      family_name: family_name,
      sub: sub
    } do
      cognito_input = %{
        UserPoolId: "user_pool_id",
        Username: username
      }

      cognito_output = %{
        "UserAttributes" => [
          %{"Name" => "sub", "Value" => sub},
          %{"Name" => "email_verified", "Value" => "true"},
          %{"Name" => "name", "Value" => name},
          %{"Name" => "family_name", "Value" => family_name},
          %{"Name" => "email", "Value" => username}
        ],
        "Username" => sub
      }

      expect(CognitoMock, :admin_get_user, fn cognito_input ->
        send(self(), {:admin_get_user, cognito_input})
        {:ok, cognito_output, ""}
      end)

      assert {:ok, user_attributes} = Cognitex.admin_get_user(username)

      assert %{
               user_attributes: %{
                 email: ^username,
                 email_verified: "true",
                 family_name: ^family_name,
                 name: ^name,
                 sub: ^sub
               }
             } = user_attributes

      assert_received {:admin_get_user, ^cognito_input}
    end

    test "should fail when incorrect username provided and return error", %{username: username} do
      cognito_input = %{
        UserPoolId: "user_pool_id",
        Username: username
      }

      cognito_status = "UserNotFoundException"
      cognito_message = "User does not exist."

      expect(CognitoMock, :admin_get_user, fn cognito_input ->
        send(self(), {:admin_get_user, cognito_input})
        {:error, {cognito_status, cognito_message}}
      end)

      assert {:error, %{message: ^cognito_message, status: ^cognito_status}} =
               Cognitex.admin_get_user(username)

      assert_received {:admin_get_user, ^cognito_input}
    end
  end

  describe "change_password/3" do
    setup :verify_on_exit!

    setup do
      {:ok, %{token: "token", previous_password: "Test123", proposed_password: "Test321"}}
    end

    test "should change user password with provided previous and new password", %{
      token: token,
      previous_password: previous_password,
      proposed_password: proposed_password
    } do
      cognito_input = %{
        AccessToken: token,
        PreviousPassword: previous_password,
        ProposedPassword: proposed_password
      }

      expect(CognitoMock, :change_password, fn cognito_input ->
        send(self(), {:change_password, cognito_input})
        {:ok, %{}, ""}
      end)

      assert {:ok, %{}} = Cognitex.change_password(token, previous_password, proposed_password)

      assert_received {:change_password, ^cognito_input}
    end

    test "should fail when incorrect username provided and return error", %{
      token: token,
      previous_password: previous_password,
      proposed_password: proposed_password
    } do
      cognito_input = %{
        AccessToken: token,
        PreviousPassword: previous_password,
        ProposedPassword: proposed_password
      }

      cognito_status = "NotAuthorizedException"
      cognito_message = "Incorrect username or password."

      expect(CognitoMock, :change_password, fn cognito_input ->
        send(self(), {:change_password, cognito_input})
        {:error, {cognito_status, cognito_message}}
      end)

      assert {:error, %{message: ^cognito_message, status: ^cognito_status}} =
               Cognitex.change_password(token, previous_password, proposed_password)

      assert_received {:change_password, ^cognito_input}
    end
  end

  describe "update_user_attributes/2" do
    setup :verify_on_exit!

    setup do
      {:ok, %{token: "token", attrs: [name: "John", family_name: "Smith"]}}
    end

    test "should update user attributes with provided data", %{
      token: token,
      attrs: [name: name, family_name: family_name] = attrs
    } do
      cognito_input = %{
        AccessToken: token,
        UserAttributes: [
          %{Name: "name", Value: name},
          %{Name: "family_name", Value: family_name}
        ]
      }

      expect(CognitoMock, :update_user_attributes, fn cognito_input ->
        send(self(), {:update_user_attributes, cognito_input})
        {:ok, %{}, ""}
      end)

      assert {:ok, %{}} = Cognitex.update_user_attributes(token, attrs)

      assert_received {:update_user_attributes, ^cognito_input}
    end

    test "should fail when incorrect token provided and return error", %{
      token: token,
      attrs: [name: name, family_name: family_name] = attrs
    } do
      cognito_input = %{
        AccessToken: token,
        UserAttributes: [
          %{Name: "name", Value: name},
          %{Name: "family_name", Value: family_name}
        ]
      }

      cognito_status = "NotAuthorizedException"
      cognito_message = "Invalid Access Token"

      expect(CognitoMock, :update_user_attributes, fn cognito_input ->
        send(self(), {:update_user_attributes, cognito_input})
        {:error, {cognito_status, cognito_message}}
      end)

      assert {:error, %{message: ^cognito_message, status: ^cognito_status}} =
               Cognitex.update_user_attributes(token, attrs)

      assert_received {:update_user_attributes, ^cognito_input}
    end
  end

  describe "forgot_password/1" do
    setup :verify_on_exit!

    setup do
      {:ok, %{username: "john.smith@example.com"}}
    end

    test "should request forgot password with provided username", %{username: username} do
      cognito_input = %{
        ClientId: "client_id",
        Username: username
      }

      cognito_output = %{
        "CodeDeliveryDetails" => %{
          "AttributeName" => "email",
          "DeliveryMedium" => "EMAIL",
          "Destination" => "j***@e***.co"
        }
      }

      expect(CognitoMock, :forgot_password, fn cognito_input ->
        send(self(), {:forgot_password, cognito_input})
        {:ok, cognito_output, ""}
      end)

      assert {:ok, ^cognito_output} = Cognitex.forgot_password(username)

      assert_received {:forgot_password, ^cognito_input}
    end

    test "should fail when provided username does not exist and return error", %{
      username: username
    } do
      cognito_input = %{
        ClientId: "client_id",
        Username: username
      }

      cognito_status = "UserNotFoundException"
      cognito_message = "Username/client id combination not found."

      expect(CognitoMock, :forgot_password, fn cognito_input ->
        send(self(), {:forgot_password, cognito_input})
        {:error, {cognito_status, cognito_message}}
      end)

      assert {:error, %{message: ^cognito_message, status: ^cognito_status}} =
               Cognitex.forgot_password(username)

      assert_received {:forgot_password, ^cognito_input}
    end
  end

  describe "confirm_forgot_password/3" do
    setup :verify_on_exit!

    setup do
      {:ok,
       %{username: "john.smith@example.com", confirmation_code: "123456", password: "Test456"}}
    end

    test "should request forgot password with provided username", %{
      username: username,
      confirmation_code: confirmation_code,
      password: password
    } do
      cognito_input = %{
        ClientId: "client_id",
        ConfirmationCode: confirmation_code,
        Password: password,
        Username: username
      }

      cognito_output = %{
        "CodeDeliveryDetails" => %{
          "AttributeName" => "email",
          "DeliveryMedium" => "EMAIL",
          "Destination" => "j***@e***.co"
        }
      }

      expect(CognitoMock, :confirm_forgot_password, fn cognito_input ->
        send(self(), {:confirm_forgot_password, cognito_input})
        {:ok, cognito_output, ""}
      end)

      assert {:ok, ^cognito_output} =
               Cognitex.confirm_forgot_password(confirmation_code, username, password)

      assert_received {:confirm_forgot_password, ^cognito_input}
    end

    test "should fail when provided confirmation code is invalid and return error", %{
      username: username,
      confirmation_code: confirmation_code,
      password: password
    } do
      cognito_input = %{
        ClientId: "client_id",
        ConfirmationCode: confirmation_code,
        Password: password,
        Username: username
      }

      cognito_status = "ExpiredCodeException"
      cognito_message = "Invalid code provided, please request a code again."

      expect(CognitoMock, :confirm_forgot_password, fn cognito_input ->
        send(self(), {:confirm_forgot_password, cognito_input})
        {:error, {cognito_status, cognito_message}}
      end)

      assert {:error, %{message: ^cognito_message, status: ^cognito_status}} =
               Cognitex.confirm_forgot_password(confirmation_code, username, password)

      assert_received {:confirm_forgot_password, ^cognito_input}
    end
  end
end


mix.exs
defmodule Cognitex.MixProject do
  use Mix.Project

  def project do
    [
      app: :cognitex,
      version: "0.1.0",
      elixir: "~> 1.7.2",
      start_permanent: Mix.env() == :prod,
      description: description(),
      package: package(),
      deps: deps()
    ]
  end

  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp description do
    """
    Library for managing user acounts through AWS Cognito service.
    """
  end

  defp package do
    [
      files: ["lib", "mix.exs", "README*", "LICENSE*"],
      maintainers: ["Adam Skołuda"],
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/Skoda091/cognitex"}
    ]
  end

  defp deps do
    [
      {:aws, "~> 0.5.0"},
      {:ex_doc, "~> 0.19", only: :dev, runtime: false},
      {:dialyxir, "~> 1.0.0-rc.3", only: :dev},
      {:credo, "~> 0.10.0", only: :dev, runtime: false},
      {:mox, "~> 0.4", only: :test}
    ]
  end
end


lib/services/cognito.ex
defmodule Services.Cognito do
  @moduledoc """
  Behaviour for AWS Cognito service.
  """
  @callback sign_up(map()) :: {:ok, map(), any()} | {:error, {String.t(), String.t()}}
  @callback confirm_sign_up(map()) :: {:ok, map(), any()} | {:error, {String.t(), String.t()}}
  @callback admin_initiate_auth(map()) :: {:ok, map(), any()} | {:error, {String.t(), String.t()}}
  @callback get_user(map()) :: {:ok, map(), any()} | {:error, {String.t(), String.t()}}
  @callback admin_get_user(map()) :: {:ok, map(), any()} | {:error, {String.t(), String.t()}}
  @callback change_password(map()) :: {:ok, map(), any()} | {:error, {String.t(), String.t()}}
  @callback update_user_attributes(map()) ::
              {:ok, map(), any()} | {:error, {String.t(), String.t()}}
  @callback forgot_password(map()) :: {:ok, map(), any()} | {:error, {String.t(), String.t()}}
  @callback confirm_forgot_password(map()) ::
              {:ok, map(), any()} | {:error, {String.t(), String.t()}}
end


lib/cognitex.ex
defmodule Cognitex do
  @moduledoc """
  Module for managing user acounts through AWS Cognito service.
  """
  import MapHelpers
  alias Services.CognitoWrapper

  @doc """
  Registers the user in the specified user pool and creates a user name, password, and user attributes defined in AWS Cognito service.

  ## Examples

      iex> Cognitex.sign_up("john.smith@example.com", "Test123", name: "John", family_name: "Smith")
      {:ok,
        %{
          "CodeDeliveryDetails" => %{
            "AttributeName" => "email",
            "DeliveryMedium" => "EMAIL",
            "Destination" => "j***@e***.co"
          },
          "UserConfirmed" => false,
          "UserSub" => "uuid"
        }
      }

      iex> Cognitex.sign_up("john.smith@example.com", "Test123", name: "John", family_name: "Smith")
      {:error,
        %{
          message: "An account with the given email already exists.",
          status: "UsernameExistsException"
        }
      }
  """
  @spec sign_up(String.t(), String.t(), [{atom(), String.t()}]) :: {:ok, map()} | {:error, map()}
  def sign_up(email, password, attrs) do
    input =
      %{}
      |> inject_client_id()
      |> inject_user_pool_id()
      |> inject_params(username: email, password: password)
      |> inject_user_attributes(attrs)

    case cognito().sign_up(input) do
      {:ok, request_data, _} -> {:ok, request_data}
      {:error, {status, message}} -> {:error, %{status: status, message: message}}
    end
  end

  @doc """
  Confirms registration of a user.

  ## Examples

      iex> Cognitex.confirm("john.smith@example.com", "123456")
      {:ok, %{}}

      iex> Cognitex.confirm("john.smith@example.com", "123456")
      {:error,
        %{
          message: "User cannot be confirm. Current status is CONFIRMED",
          status: "NotAuthorizedException"
        }
      }
  """
  @spec confirm(String.t(), String.t()) :: {:ok, map()} | {:error, map()}
  def confirm(email, confirmation_code) do
    input =
      %{}
      |> inject_client_id()
      |> inject_params(username: email, confirmation_code: confirmation_code)

    case cognito().confirm_sign_up(input) do
      {:ok, request_data, _} -> {:ok, request_data}
      {:error, {status, message}} -> {:error, %{status: status, message: message}}
    end
  end

  @doc """
  Authenticates registered user with credentials.

  ## Examples

      iex> Cognitex.authenticate("john.smith@example.com", "Test123")
      {:ok,
        %{
          "AuthenticationResult" => %{
            "AccessToken" => "<jwt_access_token>",
            "ExpiresIn" => 3600,
            "IdToken" => "<jwt_id_token>",
            "RefreshToken" => "<jwt_refresh_token>",
            "TokenType" => "Bearer"
          },
          "ChallengeParameters" => %{}
        }
      }

      iex> Cognitex.authenticate("john.smith@example.com", "Test123")
      {:error,
        %{
          message: "Incorrect username or password.",
          status: "NotAuthorizedException"
        }
      }
  """
  @spec authenticate(String.t(), String.t()) :: {:ok, map()} | {:error, map()}
  def authenticate(email, password) do
    input =
      %{}
      |> inject_auth_flow()
      |> inject_client_id()
      |> inject_user_pool_id()
      |> inject_auth_parameters(email, password)

    case cognito().admin_initiate_auth(input) do
      {:ok, request_data, _} -> {:ok, request_data}
      {:error, {status, message}} -> {:error, %{status: status, message: message}}
    end
  end

  @doc """
  Gets the user attributes and metadata for a user by access token.

  ## Examples

      iex> Cognitex.get_user("<jwt_access_token>")
      {:ok,
        %{
          user_attributes: %{
            email: "john.smith@example.com",
            email_verified: "true",
            family_name: "Smith",
            name: "John",
            sub: "<sub>"
          }
        }
      }

      iex> Cognitex.get_user("<jwt_access_token>")
      {:error,
        %{
          message: "Could not verify signature for Access Token",
          status: "NotAuthorizedException"
        }
      }
  """
  @spec get_user(String.t()) :: {:ok, map()} | {:error, map()}
  def get_user(token) do
    input =
      %{}
      |> inject_params(access_token: token)

    case cognito().get_user(input) do
      {:ok, request_data, _} -> {:ok, parse_user_attributes(request_data)}
      {:error, {status, message}} -> {:error, %{status: status, message: message}}
    end
  end

  @doc """
  Gets the specified user by username in a user pool as an administrator. Works on any user.

  ## Examples

      iex> Cognitex.admin_get_user("john.smith@example.com")
      {:ok,
        %{
          user_attributes: %{
            email: "john.smith@example.com",
            email_verified: "true",
            family_name: "Smith",
            name: "John",
            sub: "<sub>"
          }
        }
      }

      iex> Cognitex.admin_get_user("john.smith@example.com")
      {:error,
        %{
          message: "User does not exist.",
          status: "UserNotFoundException"
        }
      }
  """
  @spec admin_get_user(String.t()) :: {:ok, map()} | {:error, map()}
  def admin_get_user(username) do
    input =
      %{}
      |> inject_user_pool_id()
      |> inject_params(username: username)

    case cognito().admin_get_user(input) do
      {:ok, request_data, _} -> {:ok, parse_user_attributes(request_data)}
      {:error, {status, message}} -> {:error, %{status: status, message: message}}
    end
  end

  @doc """
  Changes the password for a specified user in a user pool.

  ## Examples

      iex> Cognitex.change_password("<jwt_access_token>", "Test123", "Test321")
      {:ok, %{}}

      iex> Cognitex.change_password("<jwt_access_token>", "Test123", "Test321")
      {:error,
        %{
          message: "Incorrect username or password.",
          status: "NotAuthorizedException"
        }
      }
  """
  @spec change_password(String.t(), String.t(), String.t()) :: {:ok, map()} | {:error, map()}
  def change_password(token, previous_password, proposed_password) do
    input =
      %{}
      |> inject_params(
        access_token: token,
        previous_password: previous_password,
        proposed_password: proposed_password
      )

    case cognito().change_password(input) do
      {:ok, request_data, _} -> {:ok, request_data}
      {:error, {status, message}} -> {:error, %{status: status, message: message}}
    end
  end

  @doc """
  Allows a user to update a specific attribute.

  ## Examples

      iex> Cognitex.update_user_attributes("<jwt_access_token>", name: "Jane", family_name: "Doe")
      {:ok, %{}}

      iex> Cognitex.update_user_attributes("<jwt_access_token>", name: "Jane", family_name: "Doe")
      {:error,
        %{
          message: "Invalid Access Token",
          status: "NotAuthorizedException"
        }
      }
  """
  @spec update_user_attributes(String.t(), [{atom(), String.t()}]) ::
          {:ok, map()} | {:error, map()}
  def update_user_attributes(token, attrs) do
    input =
      %{}
      |> inject_params(access_token: token)
      |> inject_user_attributes(attrs)

    case cognito().update_user_attributes(input) do
      {:ok, request_data, _} -> {:ok, request_data}
      {:error, {status, message}} -> {:error, %{status: status, message: message}}
    end
  end

  @doc """
  Calling this API causes a message to be sent to the end user with a confirmation code that is required to change the user's password.

  ## Examples

      iex> Cognitex.forgot_password("john.smith@example.com")
      {:ok,
        %{
          "CodeDeliveryDetails" => %{
            "AttributeName" => "email",
            "DeliveryMedium" => "EMAIL",
            "Destination" => "j***@e***.co"
          }
        }
      }

      iex> Cognitex.forgot_password("john.smith@example.com")
      {:error,
        %{
          message: "Username/client id combination not found.",
          status: "UserNotFoundException"
        }
      }
  """
  @spec forgot_password(String.t()) :: {:ok, map()} | {:error, map()}
  def forgot_password(username) do
    input =
      %{}
      |> inject_client_id()
      |> inject_params(username: username)

    case cognito().forgot_password(input) do
      {:ok, request_data, _} -> {:ok, request_data}
      {:error, {status, message}} -> {:error, %{status: status, message: message}}
    end
  end

  @doc """
  Allows a user to enter a confirmation code to reset a forgotten password.

  ## Examples

      iex> Cognitex.confirm_forgot_password("123456", "john.smith@example.com", "Test456")
      {:ok, %{}}

      iex> Cognitex.confirm_forgot_password("123456", "john.smith@example.com", "Test456")
      {:error,
        %{
          message: "Invalid code provided, please request a code again.",
          status: "ExpiredCodeException"
        }
      }
  """
  @spec confirm_forgot_password(String.t(), String.t(), String.t()) ::
          {:ok, map()} | {:error, map()}
  def confirm_forgot_password(code, username, password) do
    input =
      %{}
      |> inject_client_id()
      |> inject_params(confirmation_code: code, username: username, password: password)

    case cognito().confirm_forgot_password(input) do
      {:ok, request_data, _} -> {:ok, request_data}
      {:error, {status, message}} -> {:error, %{status: status, message: message}}
    end
  end

  defp inject_params(map, params) do
    {_, updated_map} =
      Enum.map_reduce(params, map, fn {key, value}, map ->
        {"", Map.put(map, atom_camelize(key), value)}
      end)

    updated_map
  end

  defp inject_user_attributes(map, attrs) do
    map = Map.put(map, :UserAttributes, [])

    {_, updated_map} =
      Enum.map_reduce(attrs, map, fn {key, value}, %{UserAttributes: user_attrs} = map ->
        {"", Map.put(map, :UserAttributes, user_attrs ++ [%{Name: to_string(key), Value: value}])}
      end)

    updated_map
  end

  defp inject_client_id(map) do
    Map.put(map, :ClientId, Application.get_env(:aws, :client_id))
  end

  defp inject_user_pool_id(map) do
    Map.put(map, :UserPoolId, Application.get_env(:aws, :user_pool_id))
  end

  defp inject_auth_flow(map) do
    Map.put(map, :AuthFlow, "ADMIN_NO_SRP_AUTH")
  end

  defp inject_auth_parameters(map, email, password) do
    Map.put(map, :AuthParameters, %{USERNAME: email, PASSWORD: password})
  end

  defp atom_camelize(atom) do
    atom |> Atom.to_string() |> Macro.camelize() |> String.to_atom()
  end

  defp parse_user_attributes(attrs) do
    parsed =
      attrs
      |> underscore_keys
      |> atomize_keys
      |> Map.get(:user_attributes)
      |> Enum.into(%{}, fn %{name: name, value: value} -> {String.to_atom(name), value} end)

    %{user_attributes: parsed}
  end

  defp cognito(), do: Application.get_env(:cognitex, :cognito, CognitoWrapper)
end
